### ISSUE TITLE
 Nightly failure in ipa server CA install (rawhide): ProviderException 

### ISSUE THREAD

**Issue by flo-renaud** [dogtagpki/pki#3560]
The nightly test `test_integration/test_fips.py::TestInstallFIPS::test_basic::setup` failed while configuring the CA server, see PR #[968](https://github.com/freeipa-pr-ci2/freeipa/pull/968) with the following [logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/77d5e39a-cb6c-11eb-9dd2-fa163ec89814/) and [report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/77d5e39a-cb6c-11eb-9dd2-fa163ec89814/report.html).

pkispawn failed while restarting pki-tomcat, with an exception reported in the [journal](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/77d5e39a-cb6c-11eb-9dd2-fa163ec89814/test_integration-test_fips.py-TestInstallFIPS-install/master.ipa.test/journal.gz):
```
Jun 12 11:09:39 master.ipa.test systemd[1]: Starting PKI Tomcat Server pki-tomcat...
Jun 12 11:09:42 master.ipa.test java[27711]: usr/lib/api/apiutil.c Could not open /run/lock/opencryptoki/LCK..APIlock
Jun 12 11:09:42 master.ipa.test pki-server[27711]: ProviderException: Could not initialize NSS
Jun 12 11:09:42 master.ipa.test systemd[1]: pki-tomcatd@pki-tomcat.service: Control process exited, code=exited, status=255/EXCEPTION
Jun 12 11:09:42 master.ipa.test systemd[1]: pki-tomcatd@pki-tomcat.service: Failed with result 'exit-code'.
```

Companion issue on freeipa side: https://pagure.io/freeipa/issue/8883

**Comment by flo-renaud** [comment]
Version installed: pki-server-10.11.0-0.2.alpha1.fc35.noarch
(see full package list [here](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/77d5e39a-cb6c-11eb-9dd2-fa163ec89814/installed_packages/installed_packages_master.log.gz))

**Comment by edewata** [comment]
Could you retry with pki-server-11.0.0-0.1.alpha1.fc35.noarch? Thanks.

**Comment by flo-renaud** [comment]
Hi @edewata 
I triggered the same test with pki-server-11.0.0-0.1.alpha1.fc35.noarch (see [installed packages](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b574fe42-d446-11eb-9fe6-fa163ec89814/installed_packages/installed_packages_master.log.gz)) but the error is still present. Full [report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b574fe42-d446-11eb-9fe6-fa163ec89814/report.html) and [logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b574fe42-d446-11eb-9fe6-fa163ec89814/): we can still see a failure when restarting pki and the same ProviderException in the journal.

**Comment by edewata** [comment]
@flo-renaud Hi, is this problem only happening in FIPS mode? Could you describe the scenario being tested here? Could you also point me to pkispawn log and CA/KRA debug log (if any) that contain the error message? Sorry, I'm not familiar with IPA CI reports/logs. Thanks.

**Comment by flo-renaud** [comment]
@edewata 
the test is not really running in fips mode, but it is faking FIPS mode (see the [code](https://pagure.io/freeipa/blob/master/f/ipatests/pytest_ipa/integration/fips.py#_8)).

In this fake FIPS mode, if you look at the [report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b574fe42-d446-11eb-9fe6-fa163ec89814/report.html), you can see that the test is calling ipa-server-install: `ipa-server-install -n ipa.test -r IPA.TEST -p Secret.123 -a Secret.123 --domain-level=1 --dirsrv-config-file /ipatests/ipatests_dse.ldif -U --setup-dns --forwarder 192.168.122.1 --auto-reverse --setup-kra` and this call fails in CA server installation.
If you move up one directory from the report link, you can access all the files collected by the test. Navigate to the file with the test name (test_integration-test_fips.py-TestInstallFIPS-install ) and the directory contains one subdirectory for each machine used in the test: client, master, replica.
Since the server failed, the interesting logs are in the [master subdirectory](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b574fe42-d446-11eb-9fe6-fa163ec89814/test_integration-test_fips.py-TestInstallFIPS-install/master.ipa.test/). From there you can navigate to var/log/ or inspect the journal. The error mentioned in this ticket is visible in the journal.

So in short, the scenario is equivalent to the installation of ipa server with a CA in FIPS mode.

**Comment by flo-renaud** [comment]
The fake userspace fips mode is achieved with the following commands:
```
echo -e "#userspace fips\n" > /etc/system-fips
mkdir -p /var/tmp/userspace-fips
echo -e "1\n" > /var/tmp/userspace-fips/fips-enabled
chcon -t sysctl_crypto_t -u system_u /var/tmp/userspace-fips/fips-enabled
mount --bind /var/tmp/userspace-fips/fips-enabled /proc/sys/crypto/fips_enabled
update-crypto-policies --set FIPS
```
then ipa-server-install is launched and fails in the pkispawn step, when pkispawn restarts pki-tomcatd.

**Comment by edewata** [comment]
@flo-renaud Thanks for the info. I'm trying to add a test case for FIPS mode:
https://github.com/edewata/pki/blob/fips/.github/workflows/ca-tests.yml#L178
https://github.com/edewata/pki/blob/fips/tests/bin/fips-enable.sh

But I got this error before I can try to install CA:
https://github.com/edewata/pki/runs/2935594889?check_suite_focus=true#step:6:12

```
Run docker exec pki ${PKIDIR}/tests/bin/fips-enable.sh
+ echo -e '#userspace fips\n'
+ mkdir -p /var/tmp/userspace-fips
+ echo -e '1\n'
+ chcon -t sysctl_crypto_t -u system_u /var/tmp/userspace-fips/fips-enabled
chcon: can't apply partial context to unlabeled file '/var/tmp/userspace-fips/fips-enabled'
```
Are these commands supposed to work in containers?

**Comment by flo-renaud** [comment]
I didn’t try myself but maybe with:
chcon —reference=/proc/sys/crypto/fips_enabled /var/tmp/userspace-fips/fips-enabled 
So that the fake file has the expected selinux context.

**Comment by edewata** [comment]
I have no idea what those commands do, so I copied them verbatim:
https://github.com/edewata/pki/blob/fips/tests/bin/fips-enable.sh

This is what I get now:
https://github.com/edewata/pki/runs/2936043215?check_suite_focus=true#step:6:12
```
Run docker exec pki ${PKIDIR}/tests/bin/fips-enable.sh
+ echo -e '#userspace fips\n'
+ mkdir -p /var/tmp/userspace-fips
+ echo -e '1\n'
+ chcon --reference=/proc/sys/crypto/fips_enabled /var/tmp/userspace-fips/fips-enabled
chcon: failed to get security context of '/proc/sys/crypto/fips_enabled': No such file or directory
```

**Comment by flo-renaud** [comment]
@edewata 
If you don’t run in selinux enforcing mode, you can skip the chcon command. It’s only used to set the proper selinux context to the file, and can be skipped in permissive mode.

**Comment by edewata** [comment]
I skipped `chcon`, now it's failing at the next command:
https://github.com/edewata/pki/runs/2936543456?check_suite_focus=true#step:6:12
```
Run docker exec pki ${PKIDIR}/tests/bin/fips-enable.sh
+ echo -e '#userspace fips\n'
+ mkdir -p /var/tmp/userspace-fips
+ echo -e '1\n'
+ mount --bind /var/tmp/userspace-fips/fips-enabled /proc/sys/crypto/fips_enabled
mount: /proc/sys/crypto/fips_enabled: mount point does not exist.
```

**Comment by flo-renaud** [comment]
@edewata I have only superficial knowledge re. containers, probably your container doesn't have /proc/sys/crypto/fips_enabled.
Can you try outside of a container?

**Comment by edewata** [comment]
@flo-renaud Could you retest with this build?
https://bodhi.fedoraproject.org/updates/FEDORA-2021-2c86bcd3f3

**Comment by flo-renaud** [comment]
@edewata the installation succeeds with the update, I added karma + a comment in bodhi

**Comment by edewata** [comment]
@flo-renaud Thanks! The build has been released to F35. I'm closing this ticket.
